pkg/agent/nodeportlocal/portcache/port_table.go

```diff
@@ -120,37 +100,49 @@ func (pt *PortTable) GetEntryByPodIPPort(ip string, port int) *NodePortData {
 }
 
 func (pt *PortTable) getFreePort(podIP string, podPort int) int {
-	pt.tableLock.Lock()
-	defer pt.tableLock.Unlock()
 	for i := pt.StartPort; i <= pt.EndPort; i++ {
 		if _, ok := pt.Table[i]; !ok {
-			pt.Table[i] = NodePortData{PodIP: podIP, PodPort: podPort}
+			socket, err := pt.LocalPortOpener.OpenLocalPort(i)
+			if err != nil {
+				continue
+			}
+			pt.Table[i] = NodePortData{
+				NodePort: i,
+				PodIP:    podIP,
+				PodPort:  podPort,
+				socket:   socket,
+			}
 			return i
 		}
 	}
 	return -1
 }
 
 func (pt *PortTable) AddRule(podIP string, podPort int) (int, error) {
-	nodeport := pt.getFreePort(podIP, podPort)
-	if nodeport < 0 {
+	pt.tableLock.Lock()
+	defer pt.tableLock.Unlock()
+	nodePort := pt.getFreePort(podIP, podPort)
+	if nodePort < 0 {
 		return 0, fmt.Errorf("no free port found")
 	}
-	err := pt.PodPortRules.AddRule(nodeport, fmt.Sprintf("%s:%d", podIP, podPort))
+	err := pt.PodPortRules.AddRule(nodePort, fmt.Sprintf("%s:%d", podIP, podPort))
```

Comment:
should we release the nodePort if `pt.PodPortRules.AddRule` fails? otherwise retry mechanism will skip `PodPortRules.AddRule`.

Reply:
Done, thanks